### PR TITLE
Refactoring DataSource tests

### PR DESCRIPTION
**Pull request by Corwinpro**
This PR approaches #53 

We refactor the `BaseTests` for `DataSources` and `GradientDataSources` to make them behave more like templates, not base classes. 

### Changes

- Most of the test methods are moved to the template, and the actual tests only need to specify inputs and outputs (most of the time)
- Refactored `setUp` method calls for each `DataSource` test
- Small refactors: we changed the way evaluation is iterating over test input and output values, to capture all test cases provided

### CODE REVIEW

**Comment by Corwinpro** [comment]
I agree that it seems too much of a hidden things going on there. My feeling is that the `TemplateTest` classes should provide the basic functionality, regardless of what a `DataSource` is doing. 

For example:

- In `TemplateTestDataSource` there are only 2 effective tests being performed: `test_basic_evaluation` verifies that the input parameters correctly match the outputs of `run`; `test_output_slots` verifies the lengths of the output slots and the actual outputs.
- `TemplateTestGradientDataSource` only adds two extra checks on the inputs and gradients consistency.

Regarding the failures of the tests - which is a question I asked myself as well - in fact, `unittest` outputs the module name, which failed, and the test name.

**Comment by Corwinpro** [comment]
Small suggestion here: We can make `TemplateTests` inherit from `object`, and the actual tests will include `unittest.TestCase` and a mixin `TemplateTests`. This resolves the visual discovery problem.

**Comment by flongford** [comment]
> Small suggestion here: We can make `TemplateTests` inherit from `object`, and the actual tests will include `unittest.TestCase` and a mixin `TemplateTests`. This resolves the visual discovery problem.

Nice! That feels like a good solution

**Comment by flongford** [comment]
So I think we might be able to come up with a compromise between explicit / implicit - each `assert...` method is able to tag on an additional traceback message as a `msg` key word argument. 

Providing a formatted string containing the source code file for the subclass that is performing the unit tests gives the developer any additional information that is lost when using the templates. This can then be built into the `TemplateTestCase`, with no need to change any subclasses.

File: itwm_example/tests/base_test_classes/base_test_data_source.py [hunk 1]
```diff
@@ -7,16 +7,28 @@
...
+class TemplateTestDataSource(unittest.TestCase):
     """ Base test class for a generic DataSource.
     """
+
     _data_source_index = None
-    test_case_values = None
-    test_case_objectives = None
+    test_inputs = []
+    test_outputs = []
     _objective_precision = 6
```
Comment (flongford) [inline itwm_example/tests/base_test_classes/base_test_data_source.py]:
```suggestion
@property
    def _test_case_traceback(self):
        """Provides a traceback string containing source file of
        subclass, formatted for detection by IDE"""

        test_case_class = type(self).__name__
        source_file = inspect.getsourcefile(type(self))
        line = inspect.getsourcelines(type(self))[1]

        return (f'\n    {test_case_class} source file:\n'
                f'{source_file}:{line}')`

```

**Comment by Corwinpro** [comment]
LGTM. Now it points to the actual `Test` class in addition to the parent method.

**Pull request merged**